tracee-rules/engine/engine_test.go

```diff
@@ -0,0 +1,250 @@
+				ParentProcessID: 1,
+				Args: []types.TraceeEventArgument{{
+					Name: "foobarbaz",
+				}},
+			},
+			inputSignature: fakeSignature{
+				getSelectedEvents: func() ([]types.SignatureEventSelector, error) {
+					return []types.SignatureEventSelector{
+						{
+							Source: "tracee",
+						},
+					}, nil
+				},
+			},
+			expectedNumEvents: 1,
+		},
+		{
+			name: "sad path - with all events selector, no source",
+			inputSignature: fakeSignature{
+				getSelectedEvents: func() ([]types.SignatureEventSelector, error) {
+					return []types.SignatureEventSelector{
+						{
+							Name: "*",
+						},
+					}, nil
+				},
+			},
+			expectedError: "signature Fake Signature doesn't declare an input source\n",
+		},
+		{
+			name: "sad path - signature init fails",
+			inputSignature: fakeSignature{
+				init: func(handler types.SignatureHandler) error {
+					return errors.New("init failed")
+				},
+			},
+			expectedNumEvents: 0,
+			expectedError:     "error initializing signature Fake Signature: init failed\n",
+		},
+		{
+			name: "sad path - getSelectedEvents returns an error",
+			inputSignature: fakeSignature{
+				getSelectedEvents: func() ([]types.SignatureEventSelector, error) {
+					return nil, errors.New("getSelectedEvents error")
+				},
+			},
+			expectedError: "error getting selected events for signature Fake Signature: getSelectedEvents error\n",
+		},
+		{
+			name: "sad path - getMetadata returns an error",
+			inputSignature: fakeSignature{
+				getMetadata: func() (types.SignatureMetadata, error) {
+					return types.SignatureMetadata{}, errors.New("getMetadata error")
+				},
+			},
+			expectedError: "error getting metadata: getMetadata error\n",
+		},
+	}
+
+	for _, tc := range testCases {
+		inputs := EventSources{}
+		inputs.Tracee = make(chan types.Event, 1)
+		outputChan := make(chan types.Finding, 1)
+		done := make(chan bool, 1)
+		var logBuf []byte
+		logger := bytes.NewBuffer(logBuf)
+
+		t.Run(tc.name, func(t *testing.T) {
+			defer func() {
+				// signal the end
+				done <- true
+
+				// cleanup
+				close(done)
+				close(outputChan)
+				close(inputs.Tracee)
+			}()
+
+			var sigs []types.Signature
+			sigs = append(sigs, &tc.inputSignature)
+
+			var gotNumEvents int
+			tc.inputSignature.onEvent = func(event types.Event) error {
+				assert.Equal(t, tc.inputEvent, event, tc.name)
+				gotNumEvents++
+				return nil
+			}
+
+			e := NewEngine(sigs, inputs, outputChan, logger)
+			go func() {
+				e.Start(done)
+			}()
+
+			// send a test event
+			e.inputs.Tracee <- tc.inputEvent
+
+			// assert
+			var gotEvent types.Event
+			time.Sleep(time.Millisecond * 1) // wait for events to propagate
+
+			if tc.expectedNumEvents <= 0 {
+				assert.Nil(t, gotEvent, tc.name)
+				assert.Zero(t, gotNumEvents, tc.name)
+			} else {
+				assert.Equal(t, tc.expectedNumEvents, gotNumEvents, tc.name)
+			}
+
+			if tc.expectedError != "" {
+				assert.Equal(t, tc.expectedError, logger.String(), tc.name)
```

Comment:
I'm not sure we should test the log output. it means that anytime anyone wants to log something, they need to update all the assertions that were made against the log. e.g: let's say that start will notify when it's starting "engine has started" to the log, this innocent change will affect tests to other functions (like this one). I also don't consider the log as a way to return the error. AFAIC there's no error returned.

Reply:
IMHO logging should be tested as it's part of the code and in certain cases (not here) include undesirable items (secrets being logged, etc.) which should be tested and guarded against. Another point would for when we use external libraries to log, they could panic in runtime. Chance of that are low happening reduce, if logging was tested.

I agree my approach isn't correct here, we should decouple casual logging from actual errors being logged or returned. I'll take it off from this PR and when we get to the point of having structured and levelled logging, re-introduce it appropriately.

Reply:
> I'm not sure we should test the log output. it means that anytime anyone wants to log something, they need to update all the assertions that were made against the log. e.g: let's say that start will notify when it's starting "engine has started" to the log, this innocent change will affect tests to other functions (like this one). I also don't consider the log as a way to return the error. AFAIC there's no error returned.

Thinking it with a fresh mind - I think there are two different cases at play here:

1. Places that just purely log and drop the error: https://github.com/aquasecurity/tracee/blob/a328618462e2b0697d291ea5c94489d0b9d7908b/tracee-rules/engine/engine.go#L76
2. Places that don't drop the error and take a conditional branch: https://github.com/aquasecurity/tracee/blob/a328618462e2b0697d291ea5c94489d0b9d7908b/tracee-rules/engine/engine.go#L98

For 1. I agree we can get away with  not testing at this time but for 2. I feel it's important to have a test for as its conditionally executed. Not testing it would mean any future logic added in the conditional branch will be untested. Thoughts?

As far as assertions go, it makes sense to loosen it a little by doing a `assert.Contains` rather than a `assert.Equal` as I have here. That way we won't be breaking tests on each new log line added that might be unrelated to the test itself as you mentioned.

Reply:
personally I don't consider this kind of log a spec'ed functionality, so I wouldn't test it because no one says what's good or bad. It's just going to be like a snapshot test which tells us if the result has changed, not if it broke. If you strongly feel it should be tested, it's fine with me, as long as it doesn't hinder development as with the scenario we described earlier

Reply:
Roger that - I handled it here: https://github.com/aquasecurity/tracee/pull/477/commits/2dfb7cacb6c9bb50e1505f5bd4dd1a6580d3ac0e